File: code/zig/10-selection_sort.zig
const std = @import("std");
const print = std.debug.print;
const time = std.time;
const Instant = time.Instant;
const allocator = std.heap.page_allocator;
const rand = std.rand;
const mem = std.mem;

/// Function to perform Selection Sort on an array.
///
/// # Parameters
///
/// - `arr`: The array to sort.
fn selectionSort(arr: []usize) void {
    const n = arr.len;
    for (0..n - 1) |i| {
        var min_index = i;

        // Find the minimum element in the unsorted portion
        for (i + 1..n) |j| {
            if (arr[j] < arr[min_index]) {
                min_index = j;
            }
        }

        // Swap the found minimum element with the first unsorted element
        mem.swap(usize, &arr[min_index], &arr[i]);
    }
}

/// Function to verify if an array is sorted.
///
/// # Parameters
///
/// - `arr`: The array to check.
fn isSorted(arr: []usize) bool {
    const n = arr.len;
    for (0..n - 1) |i| {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

/// Main function to demonstrate Selection Sort and measure execution time.
pub fn main() !void {
    const N_values = [_]usize{ 10_000, 20_000, 30_000, 40_000 };

    // Seed the random number generator
    var rng = rand.DefaultPrng.init(123); // Xoshiro256 is good enough

    for (N_values) |N| {
        // Allocate memory for the array
        var arr = try allocator.alloc(usize, N);
        defer allocator.free(arr);

        // Fill the array with random integers
        for (0..N) |i| {
            const random_value: usize = rng.next() % N;
            arr[i] = random_value;
        }

        // Measure the start time
        const start = try Instant.now();

        // Perform Selection Sort
        selectionSort(arr);

        // Measure the end time
        const end = try Instant.now();

        // Calculate the elapsed time in seconds
        const elapsed: f64 = @floatFromInt(end.since(start));
        const time_spent = elapsed / time.ns_per_s;

        // Verify that the array is sorted
        const is_sorted = isSorted(arr);

        // Print the result and time
        print("Selection Sort with N = {d}\n", .{N});
        if (is_sorted) {
            print("Array is sorted.\n", .{});
        } else {
            print("Array is NOT sorted.\n", .{});
        }
        print("Time taken: {d} seconds\n\n", .{time_spent});
    }
}

File: code/c/10-selection_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/**
 * Function to perform Selection Sort on an array.
 *
 * @param arr The array to sort.
 * @param n The size of the array.
 */
void selection_sort(int arr[], int n) {
  for (int i = 0; i < n - 1; i++) {
    int min_index = i;

    // Find the minimum element in the unsorted portion
    for (int j = i + 1; j < n; j++) {
      if (arr[j] < arr[min_index]) min_index = j;
    }

    // Swap the found minimum element with the first unsorted element
    int temp = arr[min_index];
    arr[min_index] = arr[i];
    arr[i] = temp;
  }
}

/**
 * Main function to demonstrate Selection Sort and measure execution time.
 */
int main() {
  int N_values[] = {10000, 20000, 30000, 40000};
  int num_tests = sizeof(N_values) / sizeof(N_values[0]);

  for (int t = 0; t < num_tests; t++) {
    int N = N_values[t];
    int *arr = (int *)malloc(N * sizeof(int));

    // Seed the random number generator
    srand(time(NULL));

    // Fill the array with random integers
    for (int i = 0; i < N; i++) {
      arr[i] = rand() % N;
    }

    // Measure the start time
    clock_t start_time = clock();

    // Perform Selection Sort
    selection_sort(arr, N);

    // Measure the end time
    clock_t end_time = clock();

    // Calculate the elapsed time in seconds
    double time_spent = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // Verify that the array is sorted
    int is_sorted = 1;
    for (int i = 1; i < N; i++) {
      if (arr[i - 1] > arr[i]) {
        is_sorted = 0;
        break;
      }
    }

    // Print the result and time
    printf("Selection Sort with N = %d\n", N);
    if (is_sorted)
      printf("Array is sorted.\n");
    else
      printf("Array is NOT sorted.\n");
    printf("Time taken: %f seconds\n\n", time_spent);

    // Free the allocated memory
    free(arr);
  }

  return 0;
}
